src/org/rascalmpl/library/Prelude.java

```diff
@@ -2394,24 +2394,24 @@ public INode arbNode() {
 	
 	protected final TypeReifier tr;
 
-	public IFunction parser(IValue start,  IBool allowAmbiguity, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
-	    return rascalValues.parser(start, allowAmbiguity, allowRecovery, hasSideEffects, values.bool(false), filters);
+	public IFunction parser(IValue start,  IBool allowAmbiguity, IInteger maxAmbDepth, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
+	    return rascalValues.parser(start, allowAmbiguity, maxAmbDepth, allowRecovery, hasSideEffects, values.bool(false), filters);
 	}
 
-	public IFunction parser(IValue start,  IBool allowAmbiguity, IBool hasSideEffects, ISet filters) {
-	    return rascalValues.parser(start, allowAmbiguity, values.bool(false), hasSideEffects, values.bool(false), filters);
+	public IFunction parser(IValue start,  IBool allowAmbiguity, IInteger maxAmbDepth, IBool hasSideEffects, ISet filters) {
+	    return rascalValues.parser(start, allowAmbiguity, maxAmbDepth, values.bool(false), hasSideEffects, values.bool(false), filters);
 	}
 
 	public IFunction firstAmbiguityFinder(IValue start, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
-	    return rascalValues.parser(start, values.bool(true), allowRecovery, hasSideEffects, values.bool(true), filters);
+	    return rascalValues.parser(start, values.bool(true), values.integer(-1), allowRecovery, hasSideEffects, values.bool(true), filters);
 	}
 	
-	public IFunction parsers(IValue start,  IBool allowAmbiguity, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
-        return rascalValues.parsers(start, allowAmbiguity, allowRecovery, hasSideEffects, values.bool(false), filters);
+	public IFunction parsers(IValue start,  IBool allowAmbiguity, IInteger maxAmbDepth, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
+        return rascalValues.parsers(start, allowAmbiguity, maxAmbDepth, allowRecovery, hasSideEffects, values.bool(false), filters);
     }
 
 	public IFunction firstAmbiguityFinders(IValue start, IBool allowRecovery, IBool hasSideEffects, ISet filters) {
-        return rascalValues.parsers(start, values.bool(true), allowRecovery, hasSideEffects, values.bool(true), filters);
+        return rascalValues.parsers(start, values.bool(true), values.integer(-1), allowRecovery, hasSideEffects, values.bool(true), filters);
```

Comment:
This question is for more cases, but why is in `-1` everywhere? why not `2` as the rascal code says? or `0`? 

Or for some of these, like this specific one when `allowRecovery` is `true`, you wouldn't expect the `-1` to be a good value for the allowed ambiguitiy depth?

Not to mention there are also `DefaultNodeFlattener.UNLIMITED_MAX_AMB_DEPTH` and `INodeFlattener.UNLIMITED_MAX_AMB_DEPTH`?

Reply:
This is a case of sloppy refactoring. I have corrected that now.
- To be able to turn of ambiguity filtering in Rascal, -1 is accepted as "no filtering"
- In Java code the constant `INodeFlattener.UNLIMITED_AMB_DEPTH` should be used when no ambiguity filtering should take place.
- When both `allowAmbiguities` and `allowRecovery` are set to false, the `maxAmbDepth` parameter does nothing as there will be no ambiguities in the resulting tree. In that case I use `INodeFlattener.UNLIMITED_AMB_DEPTH` as value although that is kind of arbitrary.

Reply:
Shouldn't we use the `2` case for when `allowRecovery` is true? as that is what the `ParseTree` is doing? I mean we don't want to have suddenly a very large amb forest?